Hooks for custom scripts, including example static IP configuration

grantemsley opened the pull request:
I've added a few hooks where people can insert their own scripts into the process.  This lets them keep their weird, entirely environment specific things out of the main factory script.

The included example shows how to set a static IP address before doing updates.  All the custom scripts are refreshed every time the factory runs, so you can easily modify them without creating new base images.

Resolves #11

BenjaminArmstrong commented:
I would rather have these inline in the main script with flags.  I am busy this week - so if you can refactor that would be great, but if you can't I will look at this next week.

grantemsley commented:
Are you against the idea of having hooks for custom scripts entirely?  Or just want setting static IPs to be part of the main script?

I can think of a lot of uses for those hooks - but they'd be very environment specific, not things other would be interested in - so I thought that would be a good way for people to extend the script without having to alter it.

I'll probably have time this week to refactor it into the main script.

BenjaminArmstrong commented:
Generally speaking I prefer to have fewer files (fewer files make it harder to break because of missing files).  If it can safely be added to the main file in a way that does not break other people - do that.  

On this specific item - a number of people have asked for this - so it should definitely be mainlined.

grantemsley commented:
Makes sense, especially if a lot of people will use it.  I'll remove the extra files and make the static IP part of the main script.  

I would like to still include the hooks though - they just check if a script exists and runs it if it does, so it won't break anything if there is nothing there.  It gives people an easy way to add custom action.  But I agree setting a static IP is common enough that it belongs in the script itself.

grantemsley commented:
Setting static IPs is now part of the main script, with setting variables to enable it.

BenjaminArmstrong commented on Image-Factory/Factory.ps1:
```diff
@@ -366,18 +366,51 @@ $updateCheckScriptBlock = {
     Import-Module $env:SystemDrive\Bits\PSWindowsUpdate\PSWindowsUpdate;
 
+    # Set static IP address - do not change values here, change them in FactoryVariables.ps1
+    $UseStaticIP = $false
```
This should not be here - right? Looks like it will never use a static IP with this line present.

grantemsley commented on Image-Factory/Factory.ps1:
```diff
@@ -509,7 +571,18 @@ function RunTheFactory
-            $updateCheckScriptBlock | Out-String | Out-File -FilePath "$($driveLetter):\Bits\Logon.ps1" -Width 4096;
+            if($UseStaticIP) {
+                $staticUpdateCheckScriptBlock = $updateCheckScriptBlock | Out-String
+                $staticUpdateCheckScriptBlock = $staticUpdateCheckScriptBlock.Replace('$UseStaticIP = $false', '$UseStaticIP = $true')
```
The "$UseStaticIP = $false" gets changed here if it is enabled.  In the scriptblock it is set to false so that nothing needed to be changed if static IP isn't enabled.

BenjaminArmstrong replied:
Okay, I see what you are doing here.  However, this is confusing for someone reading the code.  Rather than having '$UseStaticIP = $false' and changing it if true - can we have '$UseStaticIP = STATICIPBOOLPLACEHOLDER' and then change it to true or false here depending on the input.  It will mean a little more code - but it will make it readable.

grantemsley replied:
Good point - I did it that way so if not using static IPs, the scriptblock can just be run as is.  But I can see how it would be confusing to anyone who didn't write that.

Give me a bit and I'll fix it up.

grantemsley commented:
I basically copied the way peterschen implemented generic sysprep - taking the scriptblock as a string and replacing parts of it.

So if static IP is disabled, it remains false in the scriptblock and gets skipped.
If it is enabled, it will be rewritten to $true later on, and the placeholders replaced with the proper values.

grantemsley commented:
I've changed it to use a placeholder, hopefully that is less confusing.

Also refactored setting those placeholders into a function, so it doesn't have to be repeated multiple times.
